jax/_src/lax/lax.py

```diff
@@ -784,6 +784,59 @@ def dot_general(lhs: ArrayLike, rhs: ArrayLike, dimension_numbers: DotDimensionN
                             precision=canonicalize_precision(precision),
                             preferred_element_type=preferred_element_type)
 
+
+def ragged_dot(
+    lhs: Array,
+    rhs: Array,
+    group_sizes: Array,
+    precision: PrecisionLike = None,
+    preferred_element_type: DTypeLike | None = None,
+    group_offset: Array | None = None,
+    existing_output: Array | None = None,
+    ) -> Array:
+  """Ragged matrix multiplication.
+
+  Args:
+    lhs: (m, k) shaped array.
+    rhs: (g, k, n) shaped array.
+    group_sizes: (g,) shaped array with integer element type, where g denotes   number of groups. The ith element indicates the size of ith group.
+    precision: Optional. Consistent with precision argument for jax.lax.dot: https://jax.readthedocs.io/en/latest/_autosummary/.html.
+    preferred_element_type: Optional. Consistent with precision argument for jax.lax.dot: https://jax.readthedocs.io/en/latest/_autosummary/.html.
+    group_offset: Optional. (1,) shaped array that ndicates the group in group_sizes to start computing from. If not specified, defaults to [0].
+    existing_output: Optional. (m, n) shaped array with elements of type preferred_element_type, where the output should we written to. Defaults to None, in which case the output should be written to a newly allocated array.
+
+  Results:
+    (m, n) shaped array with preferred_element_type element type.
+  """
+  if group_offset:
+    raise NotImplementedError("Unimplemented group_offset support.")
+  if existing_output:
+    raise NotImplementedError("Unimplemented existing_output support.")
+  m, k = lhs.shape
+  group_count, rk, _ = rhs.shape
+  assert k == rk
+  assert group_count == group_sizes.shape[0]
+  shape = (group_count, m, k)
+  lhs = broadcast_in_dim(lhs, shape, [1, 2])
+  iota = broadcasted_iota(np.int32, shape, 1)
+  group_ends = np.cumsum(group_sizes)
+  group_starts = concatenate(
+      [_zeros(group_sizes)[:1], group_ends[:-1]], dimension=0,
+  )
+  group_ends = broadcast_in_dim(group_ends, shape, (0,))
+  group_starts = broadcast_in_dim(group_starts, shape, (0,))
+  mask = bitwise_and(group_starts <= iota, iota < group_ends)
+  lhs = select(mask, lhs, _zeros(lhs))
+  result = dot_general(
+      lhs,
+      rhs,
+      dimension_numbers=(([2], [1]), ([0], [0])),
+      precision=precision,
+      preferred_element_type=preferred_element_type,
+  )
+  return np.sum(result, axis=0, dtype=result.dtype)
```

Comment:
If we're using ordinary numpy and this is just a reference implementation, can you put it in lax_reference.py instead?

If this is meant to be a lax (i.e. HLO-ish) implementation, can you avoid calling ordinary numpy and just call jnp or lax functions instead?

Reply:
Given we already have a reference implementation in lax_reference.py, I'm not sure what this one is for (unless it's meant to be turned into one that doesn't call ordinary numpy).

Reply:
This is meant to be a lax implementation that is lowerable to HLO. I modified the implementation so that no numpy functions are used. 

I am quite surprised that when using numpy (np.cumsum and np.sum) in the lax.ragged_dot, it is still lowered to HLO as expected, yielding:

```
HloModule jit_wrapped_fun

cumsum.12 {
  ROOT Arg_0.13 = s32[1]{0} parameter(0)
}

_cumulative_reduction.14 {
  Arg_0.15 = s32[1]{0} parameter(0)
  ROOT call.16 = s32[1]{0} call(Arg_0.15), to_apply=cumsum.12
}

region_0.25 {
  Arg_0.26 = f32[] parameter(0)
  Arg_1.27 = f32[] parameter(1)
  ROOT add.28 = f32[] add(Arg_0.26, Arg_1.27)
}

ENTRY main.31 {
  constant.6 = s32[] constant(0)
  broadcast.7 = s32[1,5,4]{2,1,0} broadcast(constant.6), dimensions={}
  iota.10 = s32[5]{0} iota(), iota_dimension=0
  broadcast.11 = s32[1,5,4]{2,1,0} broadcast(iota.10), dimensions={1}
  compare.19 = pred[1,5,4]{2,1,0} compare(broadcast.7, broadcast.11), direction=LE
  Arg_2.3 = s32[1]{0} parameter(2)
  call.17 = s32[1]{0} call(Arg_2.3), to_apply=_cumulative_reduction.14
  broadcast.18 = s32[1,5,4]{2,1,0} broadcast(call.17), dimensions={0}
  compare.20 = pred[1,5,4]{2,1,0} compare(broadcast.11, broadcast.18), direction=LT
  and.21 = pred[1,5,4]{2,1,0} and(compare.19, compare.20)
  Arg_0.1 = bf16[5,4]{1,0} parameter(0)
  reshape.9 = bf16[1,5,4]{2,1,0} reshape(Arg_0.1)
  constant.4 = bf16[] constant(0)
  broadcast.5 = bf16[1,5,4]{2,1,0} broadcast(constant.4), dimensions={}
  select.22 = bf16[1,5,4]{2,1,0} select(and.21, reshape.9, broadcast.5)
  Arg_1.2 = bf16[1,4,3]{2,1,0} parameter(1)
  dot.23 = bf16[1,5,3]{2,1,0} dot(select.22, Arg_1.2), lhs_batch_dims={0}, lhs_contracting_dims={2}, rhs_batch_dims={0}, rhs_contracting_dims={1}
  convert.24 = f32[1,5,3]{2,1,0} convert(dot.23)
  constant.8 = f32[] constant(0)
  reduce.29 = f32[5,3]{1,0} reduce(convert.24, constant.8), dimensions={0}, to_apply=region_0.25
  ROOT convert.30 = bf16[5,3]{1,0} convert(reduce.29)
} // main.31
```

Reply:
> I am quite surprised that when using numpy (np.cumsum and np.sum) in the lax.ragged_dot, it is still lowered to HLO as expected, yielding:

Yeah interesting! Me too. I haven't read this whole PR yet, but I guess it must be that, at least in the tests you have, the numpy stuff was all trace-time static data, i.e. stuff that could be evaluated while running the Python? In that case, writing original numpy calls is like saying "I want this to be evaluated at Python tracing time, not staged out into the HLO".

But I would expect that in general the sizes involved here would be dynamic, ie not available until runtime.

Reply:
Got it, all uses of np are removed now. Thank you.